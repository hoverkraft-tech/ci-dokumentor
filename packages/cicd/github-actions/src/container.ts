import {
  Container,
  SECTION_GENERATOR_ADAPTER_IDENTIFIER,
  GENERATOR_ADAPTER_IDENTIFIER,
  MIGRATION_ADAPTER_IDENTIFIER,
 initContainer as coreInitContainer } from '@ci-dokumentor/core';
import { Container as InversifyContainer } from 'inversify';
import { initContainer as gitInitContainer } from '@ci-dokumentor/repository-git';
import { initContainer as githubInitContainer } from '@ci-dokumentor/repository-github';
import { GitHubActionsParser } from './github-actions-parser.js';
import { GitHubActionsGeneratorAdapter } from './github-actions-generator.adapter.js';

// Section generators
import { HeaderSectionGenerator } from './section/header-section-generator.adapter.js';
import { BadgesSectionGenerator } from './section/badges-section-generator.adapter.js';
import { OverviewSectionGenerator } from './section/overview-section-generator.adapter.js';
import { UsageSectionGenerator } from './section/usage-section-generator.adapter.js';
import { InputsSectionGenerator } from './section/inputs-section-generator.adapter.js';
import { OutputsSectionGenerator } from './section/outputs-section-generator.adapter.js';
import { SecretsSectionGenerator } from './section/secrets-section-generator.adapter.js';
import { ExamplesSectionGenerator } from './section/examples-section-generator.adapter.js';
import { ContributingSectionGenerator } from './section/contributing-section-generator.adapter.js';
import { SecuritySectionGenerator } from './section/security-section-generator.adapter.js';
import { LicenseSectionGenerator } from './section/license-section-generator.adapter.js';
import { GeneratedSectionGenerator } from './section/generated-section-generator.adapter.js';

// Migration adapters
import { ActionDocsMigrationAdapter } from './migration/action-docs-migration.adapter.js';
import { AutoDocMigrationAdapter } from './migration/auto-doc-migration.adapter.js';
import { ActdocsMigrationAdapter } from './migration/actdocs-migration.adapter.js';
import { GitHubActionReadmeGeneratorMigrationAdapter } from './migration/github-action-readme-generator-migration.adapter.js';

let container: Container | null = null;

export function resetContainer(): void {
  container = null;
}

export function initContainer(
  baseContainer: Container | undefined = undefined
): Container {
  if (baseContainer) {
    // When a base container is provided, always use it and set it as our singleton
    container = baseContainer;
  } else if (container) {
    // Only return existing singleton if no base container is provided
    return container;
  } else {
    container = new InversifyContainer() as Container;
  }

  // Return early if already bound
  if (container.isBound(GitHubActionsParser)) {
    return container;
  }

  // Bind GitHub Actions specific services only
  // Bind parser
  container.bind(GitHubActionsParser).toSelf().inSingletonScope();

  // Bind generator
  container.bind(GitHubActionsGeneratorAdapter).toSelf().inSingletonScope();
  container
    .bind(GENERATOR_ADAPTER_IDENTIFIER)
    .to(GitHubActionsGeneratorAdapter);

  // Bind section generators
  container
    .bind(SECTION_GENERATOR_ADAPTER_IDENTIFIER)
    .to(HeaderSectionGenerator);
  container
    .bind(SECTION_GENERATOR_ADAPTER_IDENTIFIER)
    .to(BadgesSectionGenerator);
  container
    .bind(SECTION_GENERATOR_ADAPTER_IDENTIFIER)
    .to(OverviewSectionGenerator);
  container
    .bind(SECTION_GENERATOR_ADAPTER_IDENTIFIER)
    .to(UsageSectionGenerator);
  container
    .bind(SECTION_GENERATOR_ADAPTER_IDENTIFIER)
    .to(InputsSectionGenerator);
  container
    .bind(SECTION_GENERATOR_ADAPTER_IDENTIFIER)
    .to(OutputsSectionGenerator);
  container
    .bind(SECTION_GENERATOR_ADAPTER_IDENTIFIER)
    .to(SecretsSectionGenerator);
  container
    .bind(SECTION_GENERATOR_ADAPTER_IDENTIFIER)
    .to(ExamplesSectionGenerator);
  container
    .bind(SECTION_GENERATOR_ADAPTER_IDENTIFIER)
    .to(ContributingSectionGenerator);
  container
    .bind(SECTION_GENERATOR_ADAPTER_IDENTIFIER)
    .to(SecuritySectionGenerator);
  container
    .bind(SECTION_GENERATOR_ADAPTER_IDENTIFIER)
    .to(LicenseSectionGenerator);
  container
    .bind(SECTION_GENERATOR_ADAPTER_IDENTIFIER)
    .to(GeneratedSectionGenerator);

  // Bind migration adapters to the container
  container.bind(ActionDocsMigrationAdapter).toSelf().inSingletonScope();
  container
    .bind(MIGRATION_ADAPTER_IDENTIFIER)
    .to(ActionDocsMigrationAdapter);

  container.bind(AutoDocMigrationAdapter).toSelf().inSingletonScope();
  container
    .bind(MIGRATION_ADAPTER_IDENTIFIER)
    .to(AutoDocMigrationAdapter);

  container.bind(ActdocsMigrationAdapter).toSelf().inSingletonScope();
  container
    .bind(MIGRATION_ADAPTER_IDENTIFIER)
    .to(ActdocsMigrationAdapter);

  container.bind(GitHubActionReadmeGeneratorMigrationAdapter).toSelf().inSingletonScope();
  container
    .bind(MIGRATION_ADAPTER_IDENTIFIER)
    .to(GitHubActionReadmeGeneratorMigrationAdapter);

  return container;
}

/**
 * Initialize a test container for GitHub Actions package.
 * This initializes all the proper packages needed for GitHub Actions testing.
 */
export function initTestContainer(): Container {
  // Initialize core container first
  const baseContainer = coreInitContainer();

  // Initialize git repository package
  gitInitContainer(baseContainer);

  // Initialize github repository package
  githubInitContainer(baseContainer);

  // Initialize this package
  return initContainer(baseContainer);
}
