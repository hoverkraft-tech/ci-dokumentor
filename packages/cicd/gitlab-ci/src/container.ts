import {
  Container,
  GENERATOR_ADAPTER_IDENTIFIER,
  initContainer as coreInitContainer
} from '@ci-dokumentor/core';
import { Container as InversifyContainer } from 'inversify';
import { initContainer as gitInitContainer } from '@ci-dokumentor/repository-git';
import { initContainer as gitlabInitContainer } from '@ci-dokumentor/repository-gitlab';
import { GitLabCIParser } from './gitlab-ci-parser.js';
import { 
  GitLabCIGeneratorAdapter, 
  GITLAB_CI_SECTION_GENERATOR_ADAPTER_IDENTIFIER 
} from './gitlab-ci-generator.adapter.js';

// Section generators
import { HeaderSectionGenerator } from './section/header-section-generator.adapter.js';
import { OverviewSectionGenerator } from './section/overview-section-generator.adapter.js';
import { UsageSectionGenerator } from './section/usage-section-generator.adapter.js';
import { InputsSectionGenerator } from './section/inputs-section-generator.adapter.js';

let container: Container | null = null;

export function resetContainer(): void {
  container = null;
}

export function initContainer(
  baseContainer: Container | undefined = undefined
): Container {
  if (baseContainer) {
    // When a base container is provided, always use it and set it as our singleton
    container = baseContainer;
  } else if (container) {
    // Only return existing singleton if no base container is provided
    return container;
  } else {
    container = new InversifyContainer() as Container;
  }

  // Return early if already bound
  if (container.isBound(GitLabCIParser)) {
    return container;
  }

  // Bind GitLab CI specific services only
  // Bind parser
  container.bind(GitLabCIParser).toSelf().inSingletonScope();

  // Bind generator
  container.bind(GitLabCIGeneratorAdapter).toSelf().inSingletonScope();
  container
    .bind(GENERATOR_ADAPTER_IDENTIFIER)
    .to(GitLabCIGeneratorAdapter);

  // Bind section generators
  container
    .bind(GITLAB_CI_SECTION_GENERATOR_ADAPTER_IDENTIFIER)
    .to(HeaderSectionGenerator);
  container
    .bind(GITLAB_CI_SECTION_GENERATOR_ADAPTER_IDENTIFIER)
    .to(OverviewSectionGenerator);
  container
    .bind(GITLAB_CI_SECTION_GENERATOR_ADAPTER_IDENTIFIER)
    .to(UsageSectionGenerator);
  container
    .bind(GITLAB_CI_SECTION_GENERATOR_ADAPTER_IDENTIFIER)
    .to(InputsSectionGenerator);

  return container;
}

/**
 * Initialize a test container for GitLab CI package.
 * This initializes all the proper packages needed for GitLab CI testing.
 */
export function initTestContainer(): Container {
  // Initialize core container first
  const baseContainer = coreInitContainer();

  // Initialize git repository package
  gitInitContainer(baseContainer);

  // Initialize gitlab repository package
  gitlabInitContainer(baseContainer);

  // Initialize this package
  return initContainer(baseContainer);
}