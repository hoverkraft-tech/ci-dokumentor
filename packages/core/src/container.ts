import { Container as InversifyContainer } from 'inversify';
import { GeneratorService } from './generator/generator.service.js';
import { FormatterService } from './formatter/formatter.service.js';
import { MarkdownFormatterAdapter } from './formatter/markdown/markdown-formatter.adapter.js';
import { FORMATTER_ADAPTER_IDENTIFIER } from './formatter/formatter.adapter.js';
import { RepositoryService } from './repository/repository.service.js';
import { LicenseService } from './license/license.service.js';
import { FileRendererAdapter } from './renderer/file-renderer.adapter.js';
import { DiffRendererAdapter } from './renderer/diff-renderer.adapter.js';
import { FileReaderAdapter } from './reader/file-reader.adapter.js';
import { VersionService } from './version/version.service.js';
import { MigrationService } from './migration/migration.service.js';
import { MarkdownTableGenerator } from './formatter/markdown/markdown-table.generator.js';
import { MarkdownLinkGenerator } from './formatter/markdown/markdown-link.generator.js';

export type Container = InversifyContainer;

let container: Container | null = null;

export function resetContainer(): void {
  container = null;
}

export function initContainer(
  baseContainer: Container | undefined = undefined
): Container {
  if (baseContainer) {
    // When a base container is provided, always use it and set it as our singleton
    container = baseContainer;
  } else if (container) {
    // Only return existing singleton if no base container is provided
    return container;
  } else {
    container = new InversifyContainer();
  }

  // Return early if services are already bound
  if (container.isBound(FormatterService)) {
    return container;
  }

  // Bind core services only - no dependencies on other packages
  container.bind(FormatterService).toSelf().inSingletonScope();
  container.bind(GeneratorService).toSelf().inSingletonScope();
  container.bind(FileReaderAdapter).toSelf().inSingletonScope();
  container.bind(FileRendererAdapter).toSelf().inSingletonScope();
  container.bind(DiffRendererAdapter).toSelf().inSingletonScope();
  container.bind(RepositoryService).toSelf().inSingletonScope();
  container.bind(LicenseService).toSelf().inSingletonScope();
  container.bind(VersionService).toSelf().inSingletonScope();
  container.bind(MigrationService).toSelf().inSingletonScope();

  // Formatter adapters
  container.bind(MarkdownLinkGenerator).toSelf().inSingletonScope();
  container.bind(MarkdownTableGenerator).toSelf().inSingletonScope();
  container.bind(MarkdownFormatterAdapter).toSelf().inSingletonScope();
  container
    .bind<MarkdownFormatterAdapter>(FORMATTER_ADAPTER_IDENTIFIER)
    .to(MarkdownFormatterAdapter);

  return container;
}
