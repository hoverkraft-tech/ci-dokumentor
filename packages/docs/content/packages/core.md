---
sidebar_position: 1
---

# Core Package

The `@ci-dokumentor/core` package contains the fundamental building blocks and abstractions that power the entire CI Dokumentor ecosystem.

## Overview

The core package provides:

- **Services** - Core business logic and orchestration
- **Adapters** - Interface implementations for various platforms
- **Interfaces** - Contracts for extensibility
- **Dependencies** - Dependency injection container setup

## Key Components

### Services

#### RepositoryService

Manages repository information and platform detection:

```typescript
class RepositoryService {
  getSupportedRepositoryPlatforms(): string[];
  autoDetectRepositoryPlatform(): Promise<string | null>;
  getRepository(): Promise<Repository>;
}
```

**Repository Model:**

```typescript
type Repository = {
  owner: string;
  name: string;
  url: string;
  fullName: string; // owner/name format
  logo?: string;
  license?: {
    name: string;
    spdxId: string | null;
    url: string | null;
  };
};
```

#### GeneratorService

Manages CI/CD platform adapters and documentation generation:

```typescript
class GeneratorService {
  getSupportedCicdPlatforms(): string[];
  getGeneratorAdapterByPlatform(platform: string): GeneratorAdapter | undefined;
  getSupportedSectionsForPlatform(platform: string): string[];
  autoDetectCicdPlatform(source: string): string | null;
  autoDetectCicdAdapter(source: string): GeneratorAdapter | null;
  generateDocumentationForPlatform(
    adapter: GeneratorAdapter,
    source: string,
    output?: string,
  ): Promise<string>;
}
```

#### FormatterService

Handles content formatting and output adapters:

```typescript
class FormatterService {
  getFormatterAdapterForFile(filePath: string): FormatterAdapter;
  // Additional formatting utilities
}
```

### Interfaces

#### RepositoryProvider

Interface for repository platform providers:

```typescript
interface RepositoryProvider {
  getPlatformName(): string;
  supports(): Promise<boolean>;
  getRepository(): Promise<Repository>;
}
```

#### GeneratorAdapter

Interface for CI/CD platform adapters:

```typescript
interface GeneratorAdapter {
  getPlatformName(): string;
  getSupportedSections(): string[];
  supportsSource(source: string): boolean;
  getDocumentationPath(source: string): string;
  generateDocumentation(
    source: string,
    formatterAdapter: FormatterAdapter,
    outputAdapter: OutputAdapter,
  ): Promise<void>;
}
```

#### FormatterAdapter

Interface for content formatters:

```typescript
interface FormatterAdapter {
  // Formatting methods for different content types
}
```

#### OutputAdapter

Interface for output handling:

```typescript
interface OutputAdapter {
  // Output methods for writing generated content
}
```

### Adapters

#### MarkdownFormatterAdapter

Provides Markdown-specific formatting capabilities.

#### FileOutputAdapter

Handles file system operations for documentation output.

## Dependency Injection

The core package uses [InversifyJS](https://inversify.io/) for dependency injection. The container is set up in `container.ts` and provides bindings for all services and adapters.

## Usage

The core package is used by other packages in the CI Dokumentor ecosystem:

- **CLI Package** - Uses `RepositoryService` and `GeneratorService` for command execution
- **Repository Packages** - Implement `RepositoryProvider` interface
- **CI/CD Packages** - Implement `GeneratorAdapter` interface

## Building

```bash
# Build the core package
nx build core

# Run tests
nx test core

# Run linting
nx lint core
```

### Testing

Testing for the core package is covered by the centralized developer testing guide: `../developers/testing` (see `packages/docs/content/developers/testing.md`). For package-specific tests, open the spec files under `packages/core/src/**/*.spec.ts`.

## Related Packages

- [CLI Package](./cli) - Command-line interface built on core
- [Repository Git](./repository-git) - Git repository provider implementation
- [Repository GitHub](./repository-github) - GitHub-specific repository provider
- [CI/CD GitHub Actions](./cicd-github-actions) - GitHub Actions manifest parser and generator
