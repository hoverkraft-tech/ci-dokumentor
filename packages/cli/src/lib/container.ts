import 'reflect-metadata';
import { Command as CommanderCommand } from 'commander';
import { type Container, initContainer as coreInitContainer } from '@ci-dokumentor/core';
import { initContainer as gitHubActionsInitContainer } from '@ci-dokumentor/cicd-github-actions';
import { COMMAND_IDENTIFIER, type Command } from './interfaces/command.interface.js';
import { LOGGER_IDENTIFIER, type Logger } from './interfaces/logger.interface.js';
import { PACKAGE_SERVICE_IDENTIFIER, type PackageService } from './interfaces/package-service.interface.js';
import { PROGRAM_IDENTIFIER, type Program } from './interfaces/program.interface.js';
import { CliApplication } from './application/cli-application.js';
import { GenerateCommand } from './commands/generate-command.js';
import { GenerateDocumentationUseCase } from './usecases/generate-documentation.usecase.js';
import { ConsoleLogger } from './services/console-logger.service.js';
import { FilePackageService } from './services/file-package.service.js';

let container: Container | null = null;

/**
 * Resets the container singleton for testing purposes
 */
export function resetContainer(): void {
    container = null;
}

/**
 * Creates and configures the dependency injection container
 */
export function initContainer(): Container {
    if (container) {
        return container;
    }

    // Initialize the core container
    // This allows us to use the core services and types
    container = coreInitContainer();

    // Initialize GitHub Actions specific bindings
    gitHubActionsInitContainer(container);

    // Bind core services
    container.bind<Logger>(LOGGER_IDENTIFIER).to(ConsoleLogger).inSingletonScope();
    container.bind<PackageService>(PACKAGE_SERVICE_IDENTIFIER).to(FilePackageService).inSingletonScope();
    container.bind<Program>(PROGRAM_IDENTIFIER).toConstantValue(new CommanderCommand());


    // Bind use cases
    container.bind(GenerateDocumentationUseCase).toSelf();

    // Bind commands (using multiInject pattern)
    container.bind<Command>(COMMAND_IDENTIFIER).to(GenerateCommand).inSingletonScope();

    // Bind the main application
    container.bind(CliApplication).toSelf().inSingletonScope();

    return container;
}