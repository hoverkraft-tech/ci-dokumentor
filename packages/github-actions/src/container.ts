import { Container, SECTION_GENERATOR_ADAPTER_IDENTIFIER, GENERATOR_ADAPTER_IDENTIFIER, initContainer as coreInitContainer } from '@ci-dokumentor/core';
import { GitHubActionsParser } from './github-actions-parser.js';
import { GitHubActionsGeneratorAdapter } from './github-actions-generator.adapter.js';

// Section generators
import { HeaderSectionGenerator } from './section/header-section-generator.adapter.js';
import { BadgesSectionGenerator } from './section/badges-section-generator.adapter.js';
import { OverviewSectionGenerator } from './section/overview-section-generator.adapter.js';
import { ContentsSectionGenerator } from './section/contents-section-generator.adapter.js';
import { UsageSectionGenerator } from './section/usage-section-generator.adapter.js';
import { InputsSectionGenerator } from './section/inputs-section-generator.adapter.js';
import { OutputsSectionGenerator } from './section/outputs-section-generator.adapter.js';
import { SecretsSectionGenerator } from './section/secrets-section-generator.adapter.js';
import { ExamplesSectionGenerator } from './section/examples-section-generator.adapter.js';
import { JobsSectionGenerator } from './section/jobs-section-generator.adapter.js';
import { ContributingSectionGenerator } from './section/contributing-section-generator.adapter.js';
import { SecuritySectionGenerator } from './section/security-section-generator.adapter.js';
import { LicenseSectionGenerator } from './section/license-section-generator.adapter.js';
import { GitHubRepositoryService } from './repository/github-repository.service.js';

let container: Container | null = null;

export function initContainer(baseContainer: Container | undefined = undefined): Container {
    if (container) {
        return container;
    }

    if (baseContainer) {
        container = baseContainer;
    } else {
        container = coreInitContainer();
    }

    // Services
    container.bind(GitHubRepositoryService).toSelf().inSingletonScope();

    // Bind parser
    container.bind(GitHubActionsParser).toSelf().inSingletonScope();

    // Bind generator
    container.bind(GitHubActionsGeneratorAdapter).toSelf().inSingletonScope();
    container.bind(GENERATOR_ADAPTER_IDENTIFIER).to(GitHubActionsGeneratorAdapter);

    // Bind section generators
    container.bind(SECTION_GENERATOR_ADAPTER_IDENTIFIER).to(HeaderSectionGenerator);
    container.bind(SECTION_GENERATOR_ADAPTER_IDENTIFIER).to(BadgesSectionGenerator);
    container.bind(SECTION_GENERATOR_ADAPTER_IDENTIFIER).to(OverviewSectionGenerator);
    container.bind(SECTION_GENERATOR_ADAPTER_IDENTIFIER).to(ContentsSectionGenerator);
    container.bind(SECTION_GENERATOR_ADAPTER_IDENTIFIER).to(UsageSectionGenerator);
    container.bind(SECTION_GENERATOR_ADAPTER_IDENTIFIER).to(InputsSectionGenerator);
    container.bind(SECTION_GENERATOR_ADAPTER_IDENTIFIER).to(OutputsSectionGenerator);
    container.bind(SECTION_GENERATOR_ADAPTER_IDENTIFIER).to(SecretsSectionGenerator);
    container.bind(SECTION_GENERATOR_ADAPTER_IDENTIFIER).to(ExamplesSectionGenerator);
    container.bind(SECTION_GENERATOR_ADAPTER_IDENTIFIER).to(JobsSectionGenerator);
    container.bind(SECTION_GENERATOR_ADAPTER_IDENTIFIER).to(ContributingSectionGenerator);
    container.bind(SECTION_GENERATOR_ADAPTER_IDENTIFIER).to(SecuritySectionGenerator);
    container.bind(SECTION_GENERATOR_ADAPTER_IDENTIFIER).to(LicenseSectionGenerator);

    return container;
}