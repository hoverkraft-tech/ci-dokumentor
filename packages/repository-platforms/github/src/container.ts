import { Container, initContainer as coreInitContainer } from '@ci-dokumentor/core';
import { GitHubRepositoryService } from './github-repository.service.js';

let container: Container | null = null;

export function initContainer(baseContainer: Container | undefined = undefined): Container {
    if (container) {
        return container;
    }

    if (baseContainer) {
        container = baseContainer;
    } else {
        container = coreInitContainer();
    }

    // Services
    container.bind(GitHubRepositoryService).toSelf().inSingletonScope();

    return container;
}