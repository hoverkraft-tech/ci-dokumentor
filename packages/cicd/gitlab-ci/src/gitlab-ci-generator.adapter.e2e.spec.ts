import { existsSync, readFileSync } from 'node:fs';
import { join } from 'node:path';
import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import {
  FileRendererAdapter,
  RepositoryProvider,
  RendererAdapter,
  MarkdownFormatterAdapter,
  FormatterAdapter,
} from '@ci-dokumentor/core';
import { GitRepositoryProvider } from '@ci-dokumentor/repository-git';
import { sanitizeSnapshotContent } from '@ci-dokumentor/core/tests';
import mockFs, { restore } from 'mock-fs';
import { initTestContainer } from './container.js';
import { GitLabCIGeneratorAdapter } from './gitlab-ci-generator.adapter.js';

const rootPath = join(__dirname, '../../../..');

describe('GitLabCIGeneratorAdapter', () => {
  let repositoryProvider: RepositoryProvider;
  let formatterAdapter: FormatterAdapter;
  let rendererAdapter: RendererAdapter;
  let gitLabCIGeneratorAdapter: GitLabCIGeneratorAdapter;

  beforeEach(async () => {
    // Use real dependencies from the container
    const container = initTestContainer();

    repositoryProvider = container.get(GitRepositoryProvider);
    formatterAdapter = container.get(MarkdownFormatterAdapter);
    rendererAdapter = container.get(FileRendererAdapter);

    gitLabCIGeneratorAdapter = container.get(GitLabCIGeneratorAdapter);
  });

  afterEach(() => {
    // Restore real file system
    restore();
    // Restore all mocks
    vi.restoreAllMocks();
  });

  describe('getPlatformName', () => {
    it('should return "gitlab-ci" as platform name', () => {
      // Act
      const result = gitLabCIGeneratorAdapter.getPlatformName();

      // Assert
      expect(result).toBe('gitlab-ci');
    });
  });

  describe('supportsSource', () => {
    it.each([
      'templates/test/template.yml',
      'templates/test/template.yaml',
      'templates/test.yml',
      'templates/test.yaml',
    ])('should correctly identify supported GitLab Component files', (file) => {

      expect(gitLabCIGeneratorAdapter.supportsSource(file)).toBe(true);
    });

    it.each([
      '.gitlab-ci.yml',
      '.gitlab-ci.yaml',
    ])('should correctly identify supported GitLab CI files', (file) => {
      expect(gitLabCIGeneratorAdapter.supportsSource(file)).toBe(true);
    });

    it.each([
      'package.json',
      'README.md',
      'config.yml',
      'action.yml',
      '.github/workflows/ci.yml',
      'docker-compose.yml',
    ])('should reject unsupported files', (file) => {
      expect(gitLabCIGeneratorAdapter.supportsSource(file)).toBe(false);
    });
  });

  describe('getDocumentationPath', () => {
    it.each([
      { input: 'templates/test/template.yml', expected: 'templates/test/docs.md' },
      { input: 'templates/test/template.yaml', expected: 'templates/test/docs.md' },
      { input: 'templates/test.yml', expected: 'templates/test.md' },
      { input: 'templates/test.yaml', expected: 'templates/test.md' },
    ])('should generate correct paths for GitLab Components', ({ input, expected }) => {
      expect(gitLabCIGeneratorAdapter.getDocumentationPath(input)).toBe(
        expected
      );
    });

    it.each([
      {
        input: '.gitlab-ci.yml',
        expected: '.gitlab-ci.md',
      },
      {
        input: '.gitlab-ci.yaml',
        expected: '.gitlab-ci.md',
      },
    ])('should generate correct paths for GitLab CI files', ({ input, expected }) => {
      expect(gitLabCIGeneratorAdapter.getDocumentationPath(input)).toBe(
        expected
      );
    });
  });

  describe('generateDocumentation', () => {
    it('should generate complete documentation for a GitLab Component', async () => {
      // Arrange
      const componentYaml = `
# Docker Build Component
# A reusable component for building and pushing Docker images

spec:
  inputs:
    image_name:
      description: 'Name of the Docker image to build'
      type: string
      required: true
    platforms:
      description: 'Target platforms for multi-arch builds'
      type: string
      default: 'linux/amd64,linux/arm64'
    dockerfile_path:
      description: 'Path to the Dockerfile'
      type: string
      default: 'Dockerfile'
    build_context:
      description: 'Build context directory'
      type: string
      default: '.'

docker-build:
  stage: build
  image: docker:24-dind
  services:
    - docker:24-dind
  variables:
    DOCKER_TLS_CERTDIR: "/certs"
  before_script:
    - docker info
    - echo $CI_REGISTRY_PASSWORD | docker login -u $CI_REGISTRY_USER --password-stdin $CI_REGISTRY
  script:
    - |
      docker buildx create --use --driver docker-container
      docker buildx build \\
        --platform $[[ inputs.platforms ]] \\
        --file $[[ inputs.dockerfile_path ]] \\
        --tag $[[ inputs.image_name ]] \\
        --push \\
        $[[ inputs.build_context ]]
  rules:
    - if: $CI_COMMIT_BRANCH == $CI_DEFAULT_BRANCH
    - if: $CI_PIPELINE_SOURCE == "merge_request_event"
`;

      const sourcePath = join(rootPath, 'templates/docker-build/template.yml');
      const destinationPath = join(rootPath, 'templates/docker-build/docs.md');

      // Setup mock file with the component YAML content
      mockFs({
        [sourcePath]: componentYaml,
      });

      // Act
      await rendererAdapter.initialize(
        destinationPath,
        formatterAdapter,
      );

      await gitLabCIGeneratorAdapter.generateDocumentation({
        source: sourcePath,
        sections: {},
        rendererAdapter,
        repositoryProvider,
      });

      await rendererAdapter.finalize();

      // Assert
      // Verify that the docs.md file was created and has content
      expect(existsSync(destinationPath)).toBe(true);

      const generatedContent = readFileSync(destinationPath, 'utf-8');
      expect(generatedContent).toBeDefined();
      expect(generatedContent.length).toBeGreaterThan(0);

      // Snapshot test the generated content      
      const sanitizedContent = sanitizeSnapshotContent(generatedContent);
      expect(sanitizedContent).toMatchSnapshot('gitlab-component-documentation');
    });

    it('should generate documentation for a GitLab CI Pipeline', async () => {
      // Arrange
      const pipelineYaml = `
# GitLab CI Pipeline for Node.js Application
# This pipeline builds, tests, and deploys a Node.js application

stages:
  - build
  - test
  - deploy

variables:
  NODE_VERSION: "18"
  REGISTRY: $CI_REGISTRY

workflow:
  rules:
    - if: $CI_COMMIT_BRANCH == $CI_DEFAULT_BRANCH
    - if: $CI_PIPELINE_SOURCE == "merge_request_event"
    - if: $CI_COMMIT_TAG

.node_template: &node_template
  image: node:\${NODE_VERSION}
  cache:
    paths:
      - node_modules/
  before_script:
    - npm ci

build:
  <<: *node_template
  stage: build
  script:
    - npm run build
  artifacts:
    paths:
      - dist/
    expire_in: 1 hour
  rules:
    - if: $CI_COMMIT_BRANCH == $CI_DEFAULT_BRANCH
    - if: $CI_PIPELINE_SOURCE == "merge_request_event"

test:
  <<: *node_template
  stage: test
  script:
    - npm run test:coverage
  coverage: '/Lines\\s*:\\s*(\\d+\\.?\\d*)%/'
  artifacts:
    reports:
      coverage_report:
        coverage_format: cobertura
        path: coverage/cobertura-coverage.xml
    expire_in: 1 week
  rules:
    - if: $CI_COMMIT_BRANCH == $CI_DEFAULT_BRANCH
    - if: $CI_PIPELINE_SOURCE == "merge_request_event"

lint:
  <<: *node_template
  stage: test
  script:
    - npm run lint
  rules:
    - if: $CI_COMMIT_BRANCH == $CI_DEFAULT_BRANCH
    - if: $CI_PIPELINE_SOURCE == "merge_request_event"

deploy:
  stage: deploy
  image: docker:24
  services:
    - docker:24-dind
  variables:
    DOCKER_TLS_CERTDIR: "/certs"
  before_script:
    - echo $CI_REGISTRY_PASSWORD | docker login -u $CI_REGISTRY_USER --password-stdin $CI_REGISTRY
  script:
    - docker build -t $REGISTRY/$CI_PROJECT_PATH:$CI_COMMIT_SHA .
    - docker push $REGISTRY/$CI_PROJECT_PATH:$CI_COMMIT_SHA
    - docker tag $REGISTRY/$CI_PROJECT_PATH:$CI_COMMIT_SHA $REGISTRY/$CI_PROJECT_PATH:latest
    - docker push $REGISTRY/$CI_PROJECT_PATH:latest
  environment:
    name: production
    url: https://app.example.com
  rules:
    - if: $CI_COMMIT_BRANCH == $CI_DEFAULT_BRANCH
`;
      const sourcePath = join(rootPath, '.gitlab-ci.yml');
      const destinationPath = join(rootPath, '.gitlab-ci.md');

      // Setup mock file with the pipeline YAML content
      mockFs({
        [sourcePath]: pipelineYaml,
      });

      // Act
      await rendererAdapter.initialize(
        destinationPath,
        formatterAdapter,
      );

      await gitLabCIGeneratorAdapter.generateDocumentation({
        source: sourcePath,
        sections: {},
        rendererAdapter,
        repositoryProvider,
      });

      await rendererAdapter.finalize();

      // Assert
      // Verify that the .gitlab-ci.md file was created and has content
      expect(existsSync(destinationPath)).toBe(true);

      const generatedContent = readFileSync(destinationPath, 'utf-8');
      expect(generatedContent).toBeDefined();
      expect(generatedContent.length).toBeGreaterThan(0);

      // Snapshot test the generated content
      const sanitizedContent = sanitizeSnapshotContent(generatedContent);
      expect(sanitizedContent).toMatchSnapshot('gitlab-pipeline-documentation');
    });


    it('should handle empty YAML files', async () => {
      // Arrange
      // Setup mock file with empty content
      mockFs({
        '/test/.gitlab-ci.yml': '',
      });

      // Act
      await rendererAdapter.initialize(
        '/test/.gitlab-ci.md',
        formatterAdapter,
      );

      await expect(
        gitLabCIGeneratorAdapter.generateDocumentation({
          source: '/test/.gitlab-ci.yml',
          sections: {},
          rendererAdapter,
          repositoryProvider,
        })
      ).rejects.toThrow('Unsupported source file: /test/.gitlab-ci.yml');

      await rendererAdapter.finalize();
    });
  });
});