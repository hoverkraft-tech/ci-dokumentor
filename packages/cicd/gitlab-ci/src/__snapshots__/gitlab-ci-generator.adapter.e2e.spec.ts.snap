// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GitLabCIGeneratorAdapter > generateDocumentation > should generate complete documentation for a GitLab Component > gitlab-component-documentation 1`] = `
"<!-- header:start -->

# Template

<!-- header:end -->
<!-- overview:start -->

## Overview

Docker Build Component
A reusable component for building and pushing Docker images

<!-- overview:end -->
<!-- usage:start -->

\`\`\`yaml
include:
  - component: hoverkraft-tech/ci-dokumentor@templates/docker-build/template.yml
    with:
      # Add component inputs here
\`\`\`

<!-- usage:end -->
<!-- inputs:start -->

## Inputs

| **Input**             | **Description**                        | **Required** | **Type**   | **Default**               |
| --------------------- | -------------------------------------- | ------------ | ---------- | ------------------------- |
| **\`image_name\`**      | Name of the Docker image to build      | **true**     | **string** | -                         |
| **\`platforms\`**       | Target platforms for multi-arch builds | **false**    | **string** | \`linux/amd64,linux/arm64\` |
| **\`dockerfile_path\`** | Path to the Dockerfile                 | **false**    | **string** | \`Dockerfile\`              |
| **\`build_context\`**   | Build context directory                | **false**    | **string** | \`.\`                       |

<!-- inputs:end -->
"
`;

exports[`GitLabCIGeneratorAdapter > generateDocumentation > should generate documentation for a GitLab CI Pipeline > gitlab-pipeline-documentation 1`] = `
"<!-- header:start -->

# .Gitlab Ci

<!-- header:end -->
<!-- overview:start -->

## Overview

GitLab CI Pipeline for Node.js Application
This pipeline builds, tests, and deploys a Node.js application

<!-- overview:end -->
<!-- usage:start -->

\`\`\`yaml
include:
  - project: 'hoverkraft-tech/ci-dokumentor'
    file: '.gitlab-ci.yml'
    ref: '0.1.3'
\`\`\`

<!-- usage:end -->
<!-- inputs:start -->
<!-- inputs:end -->
"
`;
